### PR TITLE
Refactor extract ebean-api dbplatform into their own platform specific module

### PR DESCRIPTION
**Pull request by rbygrave**
Need to check DB2Legacy DDL here ...

### CODE REVIEW

File: ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DefaultDbMigration.java [hunk 1]
```diff
@@ -805,58 +789,18 @@ private File modelDirectory(File migrationDirectory) {
    */
   protected DatabasePlatform platform(Platform platform) {
     switch (platform) {
-      case H2:
```
Comment (rPraml) [inline ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DefaultDbMigration.java:808]: Plattform is an enum and not easily extendable, do you think it makes sense to have a Platform.USER1 that can be used for custom implementations?

**Review by rPraml — COMMENTED** [review]
I did a short review of this...

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DatabasePlatformFactory.java [hunk 2]
```diff
@@ -71,68 +60,13 @@ public DatabasePlatform create(DatabaseConfig config) {
...
-      return new OraclePlatform();
-    }
-    if (dbName.equals("sqlserver16")) {
-      return new SqlServer16Platform();
-    }
-    if (dbName.equals("sqlserver17")) {
-      return new SqlServer17Platform();
-    }
     if (dbName.equals("sqlserver")) {
       throw new IllegalArgumentException("Please choose the more specific sqlserver16 or sqlserver17 platform. Refer to issue #1340 for details");
```
[the remark this answers is not in the document]
Comment (rbygrave) [reply ebean-core/src/main/java/io/ebeaninternal/server/core/DatabasePlatformFactory.java:64]: Yeah, looking at this again I agree, this bit should move into SqlServerPlatform.

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DatabasePlatformFactory.java [hunk 3]
```diff
@@ -156,79 +90,41 @@ private DatabasePlatform byDatabaseMeta(DatabaseMetaData metaData, Connection co
...
-      return new SQLitePlatform();
-    } else if (dbProductName.contains("db2")) {
-      throw new IllegalArgumentException("For DB2 please explicitly choose either db2legacy/db2luw/db2zos/db2fori platform. Refer to issue #2514 for details");
-    } else if (dbProductName.contains("sql anywhere")) {
-      return new SqlAnywherePlatform();
-    } else if (dbProductName.contains("hdb")) {
-      return new HanaPlatform();
-    } else if (dbProductName.contains("clickhouse")) {
-      return new ClickHousePlatform();
+      dbProductName = readPostgres(connection, productVersion);
```
[the remark this answers is not in the document]
Comment (rbygrave) [reply ebean-core/src/main/java/io/ebeaninternal/server/core/DatabasePlatformFactory.java:98]: The original intention was to move it into a provider ... and as I see it that could be done if Postgres, Cockroach and Yugabyte were all a single provider.  At the moment I've put them into 3 separate providers.

Q: Should we merge those 3 into a single provider?  ... so ebean-postgres would support postgres, cockroach, and yugabyte.  Is there benefit in keeping these as 3 separate providers?

File: ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DefaultDbMigration.java (hunk 1 above)
Comment (rbygrave) [reply ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DefaultDbMigration.java:808]: I think that is an excellent suggestion.  Perhaps Platform.CUSTOM or Platform.USERDEFINED ...

File: ebean-core/src/main/java/io/ebeaninternal/server/core/DatabasePlatformFactory.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rbygrave) [reply ebean-core/src/main/java/io/ebeaninternal/server/core/DatabasePlatformFactory.java:98]: My gut says we should merge these 3 into a single provider ... and then we can move that platform detection code into it.  My thinking is that Yugabyte and Cockroach have that dependency on ebean-postgres anyway and it's likely that projects might want to additionally test against postgres.

File: ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DefaultDbMigration.java (hunk 1 above)
Comment (rPraml) [reply ebean-ddl-generator/src/main/java/io/ebeaninternal/dbmigration/DefaultDbMigration.java:808]: +1 for CUSTOM

**Comment by rbygrave** [comment]
Replaced by https://github.com/ebean-orm/ebean/pull/2621

**Pull request closed without merging**